feat(profiling) add changed functions table

JonasBa opened the pull request:
![CleanShot 2023-12-06 at 12 18 30@2x](https://github.com/getsentry/sentry/assets/9317857/f7f0646f-b825-411c-889b-232fc7fb1e49)

JonasBa committed:
feat(profiling) add changed functions table

Zylphrex commented on static/app/components/events/eventStatisticalDetector/eventDifferentialFlamegraph.tsx:
```diff
@@ -195,6 +209,26 @@ function EventDifferentialFlamegraphView(props: EventDifferentialFlamegraphViewP
     );
   }, [beforeFlamegraph, afterFlamegraph, theme, source]);
 
+  const makeFunctionFlamechartLink = useMemo(() => {
+    return (frame: FlamegraphFrame): LocationDescriptor => {
+      if (!props.project) {
+        return '';
+      }
+      if (!frame.profileIds?.length) {
+        return '';
+      }
+      return generateProfileFlamechartRouteWithQuery({
+        orgSlug: organization.slug,
+        projectSlug: props.project.slug,
+        profileId: frame.profileIds?.[0] ?? '',
+        query: {
+          frameName: frame.frame.name,
+          framePackage: frame.frame.package,
+        },
+      });
+    };
+  }, [organization.slug, props.project]);
```
This could be a `useCallback`.

Zylphrex commented on static/app/components/events/eventStatisticalDetector/eventDifferentialFlamegraph.tsx:
```diff
@@ -316,6 +369,267 @@ function DifferentialFlamegraphTransactionToolbar(
+    case 'next':
+      return {
+        ...state,
+        page: Math.min(state.page + 1, state.pageCount),
```
Is there an off by one here? Should it be `state.pageCount - 1` since the page numbers go from `0` to `state.pageCount - 1`?

JonasBa replied:
The numbers for pageCount are 0 - pageCount inclusive as we Math.ceil on the list length / size

Zylphrex replied:
If we take 21 items with a page size of 10, then we have pages 0, 1, and 2 right? but ceiling of 21/10 = 3. So the page count is not inclusive.

Zylphrex commented on static/app/components/events/eventStatisticalDetector/eventDifferentialFlamegraph.tsx:
```diff
@@ -316,6 +369,267 @@ function DifferentialFlamegraphTransactionToolbar(
+    });
+  }, [props.functions]);
+
+  const onPreviousPaginationClick = useMemo(() => {
+    if (state.page === 0) {
+      return undefined;
+    }
+    return () => dispatch({type: 'previous'});
+  }, [state.page]);
+
+  const onNextPaginationClick = useMemo(() => {
+    if (state.page + 1 === state.pageCount) {
+      return undefined;
+    }
+    return () => dispatch({type: 'next'});
+  }, [state.page, state.pageCount]);
```
The guards here are just to disable the previous/next buttons right? As the reducer themselves already handle the events for the first/last page.

JonasBa replied:
Yup, correct!

JonasBa committed:
feat(profiling): fix flamegraph fn generation to use useCallback

JonasBa committed:
feat(profiling): pageCount is exclusive

JonasBa committed:
feat(profiling) add legend and explanation to the flamegraph (#61268)

![CleanShot 2023-12-06 at 13 07
24@2x](https://github.com/getsentry/sentry/assets/9317857/06413bb4-0a73-47cc-9b8c-3634398bed58)

---------

Co-authored-by: Tony Xiao <txiao@sentry.io>
